### ISSUE TITLE
Beacon client crashed with OOM during init-sync

### ISSUE THREAD

**Issue by torfbolt** [prysmaticlabs/prysm#4970]
Beacon client (af5cc31565629f3e766b11caf7d8a7560d6c8901) showed the following behavior (repeatedly, every time):
- Started init sync normally (with existing DB)
- After some minutes it dropped most of the peers
- After ~2min some peers were reconnected
- Then there was a sudden surge in memory usage and it crashed, presumably due to my docker memory limit of 8G

```
[2020-02-29 22:22:45]  INFO initial-sync: Processing block 362970/366713 - estimated time remaining 29m0s blocksPerSecond=2.1 peers=5/11
[2020-02-29 22:22:46]  INFO initial-sync: Processing block 362971/366713 - estimated time remaining 30m25s blocksPerSecond=2.0 peers=5/11
[2020-02-29 22:22:46]  INFO initial-sync: Processing block 362972/366713 - estimated time remaining 29m41s blocksPerSecond=2.1 peers=5/11
[2020-02-29 22:22:47]  INFO initial-sync: Processing block 362973/366713 - estimated time remaining 29m40s blocksPerSecond=2.1 peers=5/11
[2020-02-29 22:22:58]  INFO p2p: Peer disconnected active=10 peer=16Uiu2HAkwPNFxuvK9MLYNudRTUsQWJAycbde1KBy3W3Nk43PaK9x
[2020-02-29 22:23:01]  INFO p2p: Peer disconnected active=9 peer=16Uiu2HAmQSnUoRuMosvSmC7fiij2Zs7GVEvLKo8DX6diN6qr5rTg
[2020-02-29 22:23:02]  INFO p2p: Peer disconnected active=8 peer=16Uiu2HAm1kEiusD2sM9m3mqJfmsZkqB5Gtc1KeNSrXi8dzkUGw5Y
[2020-02-29 22:23:02]  INFO p2p: Peer disconnected active=7 peer=16Uiu2HAmJNZnTxJ2U3aHr8gzLb4xZnHmVXUPTNkQQMuFcP9RPg11
[2020-02-29 22:23:02]  INFO p2p: Peer disconnected active=6 peer=16Uiu2HAkzMcawe9KeF1jPXSSRdcKGtUKXV3Jwxvok4dFJ8j3MEgs
[2020-02-29 22:23:12]  INFO p2p: Peer disconnected active=7 peer=16Uiu2HAkwp8aksoPCsEaH2tFrXe4qR9vM4Yb4sH7PRncNgZWJMJm
[2020-02-29 22:23:12]  INFO p2p: Peer disconnected active=6 peer=16Uiu2HAkvPEEaGrwaof6fpjHNsxJ6GHawcCjtC7qZLsH8DRQrUKD
[2020-02-29 22:23:14]  INFO p2p: Peer connected activePeers=6 direction=2 multiAddr=/ip4/62.210.207.65/tcp/13000/p2p/16Uiu2HAmJNZnTxJ2U3aHr8gzLb4xZnHmVXUPTNkQQMuFcP9RPg11 peer=16Uiu2HAmJNZnTxJ2U3aHr8gzLb4xZnHmVXUPTNkQQMuFcP9RPg11
[2020-02-29 22:23:21]  INFO p2p: Peer disconnected active=4 peer=16Uiu2HAm9nrJVrYzqFWLMZRPqqk9H52hRpB97hx3muqCbAvHMP8h
[2020-02-29 22:23:23]  INFO p2p: Peer disconnected active=3 peer=16Uiu2HAmG5BbygmzbPe2GNhXnM5iSjFLkPoTnG19YAgm7moVQKJr
[2020-02-29 22:23:35]  INFO p2p: Peer connected activePeers=3 direction=2 multiAddr=/ip4/89.245.40.170/tcp/13000/p2p/16Uiu2HAkzMcawe9KeF1jPXSSRdcKGtUKXV3Jwxvok4dFJ8j3MEgs peer=16Uiu2HAkzMcawe9KeF1jPXSSRdcKGtUKXV3Jwxvok4dFJ8j3MEgs
[2020-02-29 22:23:36]  INFO p2p: Peer connected activePeers=4 direction=2 multiAddr=/ip4/3.234.170.62/tcp/30000/p2p/QmfAgkmjiZNZhr2wFN9TwaRgHouMTBT6HELyzE5A3BT2wK/p2p-circuit/p2p/16Uiu2HAm1kEiusD2sM9m3mqJfmsZkqB5Gtc1KeNSrXi8dzkUGw5Y peer=16Uiu2HAm1kEiusD2sM9m3mqJfmsZkqB5Gtc1KeNSrXi8dzkUGw5Y
[2020-02-29 22:24:57]  INFO p2p: Peer disconnected active=4 peer=16Uiu2HAmJNZnTxJ2U3aHr8gzLb4xZnHmVXUPTNkQQMuFcP9RPg11
[2020-02-29 22:24:57]  INFO p2p: Peer connected activePeers=4 direction=2 multiAddr=/ip4/62.210.207.65/tcp/13000/p2p/16Uiu2HAmJNZnTxJ2U3aHr8gzLb4xZnHmVXUPTNkQQMuFcP9RPg11 peer=16Uiu2HAmJNZnTxJ2U3aHr8gzLb4xZnHmVXUPTNkQQMuFcP9RPg11
[2020-02-29 22:24:59]  INFO p2p: Peer connected activePeers=5 direction=2 multiAddr=/ip4/3.234.170.62/tcp/30000/p2p/QmfAgkmjiZNZhr2wFN9TwaRgHouMTBT6HELyzE5A3BT2wK/p2p-circuit/p2p/16Uiu2HAkwp8aksoPCsEaH2tFrXe4qR9vM4Yb4sH7PRncNgZWJMJm peer=16Uiu2HAkwp8aksoPCsEaH2tFrXe4qR9vM4Yb4sH7PRncNgZWJMJm
[2020-02-29 22:24:59]  INFO p2p: Peer connected activePeers=6 direction=2 multiAddr=/ip4/188.192.41.64/tcp/13000/p2p/16Uiu2HAkwPNFxuvK9MLYNudRTUsQWJAycbde1KBy3W3Nk43PaK9x peer=16Uiu2HAkwPNFxuvK9MLYNudRTUsQWJAycbde1KBy3W3Nk43PaK9x
[2020-02-29 22:25:02]  INFO p2p: Peer connected activePeers=7 direction=2 multiAddr=/ip4/3.234.170.62/tcp/30000/p2p/QmfAgkmjiZNZhr2wFN9TwaRgHouMTBT6HELyzE5A3BT2wK/p2p-circuit/p2p/16Uiu2HAkvPEEaGrwaof6fpjHNsxJ6GHawcCjtC7qZLsH8DRQrUKD peer=16Uiu2HAkvPEEaGrwaof6fpjHNsxJ6GHawcCjtC7qZLsH8DRQrUKD
[2020-02-29 22:26:44]  INFO p2p: Peer disconnected active=7 peer=16Uiu2HAmJNZnTxJ2U3aHr8gzLb4xZnHmVXUPTNkQQMuFcP9RPg11
Beacon client crashed with exit code 137.  Respawning..
```

**Comment by prestonvanloon** [comment]
@nisdas has this been resolved?

**Comment by nisdas** [comment]
Not as far as I know of, this is a new issue it looks like.

**Comment by prestonvanloon** [comment]
Assigning to @farazdagi as owner of new initial sync work.

**Comment by farazdagi** [comment]
Synced several times using the following command (8GiB memory limit, no swap):

```bash
docker run -it -v $HOME/prysm:/data -p 4000:4000 --memory 8g --memory-swap 8g \                                                                                                                                                                                                                                    
  gcr.io/prysmaticlabs/prysm/beacon-chain:latest \
  --datadir=/data --enable-byte-mempool --enable-state-field-trie \
  --new-state-mgmt --enable-ssz-cache
```

On average, `docker stats` reported ~4GiB of memory used, but had to set limit to 8GiB as there're currently spikes (those are w/i that limit as I've been running container without the restart flag - to catch any OOMs).

**Comment by chrishobcroft** [comment]
@farazdagi I would recommend re-opening this.

I have been using 0.3.10 to run `initial-sync` for the past 4 days. The process consistently crashes after ~1.5-2 hours, simply reporting `Killed` in the output.

Upon further investigation, and measuring memory usage using `top`, the process running prysm proceeds to consume all available physical memory, then consumes all available swap, and when swap hits 0, the process fails.

Here are screenshots, charting the process:

`16:25:09` Run the script to start the process, `beacon-chain-v0`, which takes **10% of available memory**:

![Screenshot from 2020-04-10 19-35-05](https://user-images.githubusercontent.com/2212651/78997531-60d8fa00-7b64-11ea-969c-97f725a50cbe.png)

`16:27:23`, process has **20% of available memory**:

![Screenshot from 2020-04-10 19-35-19](https://user-images.githubusercontent.com/2212651/78997880-1015d100-7b65-11ea-95f3-76ff78993ce3.png)

`16:29:05`, process has **30% of available memory**:

![Screenshot from 2020-04-10 19-35-35](https://user-images.githubusercontent.com/2212651/78997943-30de2680-7b65-11ea-8107-2183e0592cdd.png)

`16:35:15`, process has **40% of available memory**:

![Screenshot from 2020-04-10 19-35-52](https://user-images.githubusercontent.com/2212651/78997966-46ebe700-7b65-11ea-83e4-8ecf9c513216.png)

`18:13:18`, process has **77.8% of available memory**, with `202.7Mb` of swap left:

![Screenshot from 2020-04-10 19-36-50](https://user-images.githubusercontent.com/2212651/78997975-4ce1c800-7b65-11ea-9cc0-d2e87819a452.png)

`18:13:24`, process rises to **78.5% of available memory**, with `60.9Mb` of swap left:

![Screenshot from 2020-04-10 20-49-26](https://user-images.githubusercontent.com/2212651/79001948-10b26580-7b6d-11ea-8ca6-cda5f92a3ff1.png)

`18:13:38`, process peaks at **79.6% of available memory**, taking all remaining swap, and fails, reporting `Killed` (`top` hasn't updated yet):

![Screenshot from 2020-04-10 19-37-06](https://user-images.githubusercontent.com/2212651/79001579-53c00900-7b6c-11ea-979c-4c33e1c0300c.png)

`after 18:13:39`, `top` updates showing `beacon-chain-v0` process is gone:

![Screenshot from 2020-04-10 20-54-26](https://user-images.githubusercontent.com/2212651/79002226-a1894100-7b6d-11ea-97a7-952acd452c2f.png)

# Environmentals:

Operating system:
```
Distributor ID:	Ubuntu
Description:	Ubuntu 19.10
Release:	19.10
Codename:	eoan
```
Hardware:
```
H/W path             Device     Class          Description
==========================================================
                                system         20NXCTO1WW (LENOVO_MT_20NX_BU_Think_FM_ThinkPad T490s)
/0                              bus            20NXCTO1WW
/0/2                            memory         8GiB System Memory
/0/2/0                          memory         4GiB SODIMM DDR4 Synchronous 2667 MHz (0.4 ns)
/0/2/1                          memory         4GiB SODIMM DDR4 Synchronous 2667 MHz (0.4 ns)
/0/c                            memory         256KiB L1 cache
/0/d                            memory         1MiB L2 cache
/0/e                            memory         6MiB L3 cache
/0/f                            processor      Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz
/0/11                           memory         128KiB BIOS
/0/100                          bridge         Intel Corporation
/0/100/2                        display        UHD Graphics 620 (Whiskey Lake)
/0/100/4                        generic        Xeon E3-1200 v5/E3-1500 v5/6th Gen Core Processor Thermal Subsystem
/0/100/8                        generic        Xeon E3-1200 v5/v6 / E3-1500 v5 / 6th/7th Gen Core Processor Gaussian Mixture Model
```
